include/clap/helpers/plugin.hxx

```diff
@@ -791,6 +800,69 @@ namespace clap { namespace helpers {
       return self.paramsCount();
    }
 
+   //--------------------------------------//
+   // clap_plugin_configurable_audio_ports //
+   //--------------------------------------//
+
+   template <MisbehaviourHandler h, CheckingLevel l>
+   bool Plugin<h, l>::clapConfigurableAudioPortsCanApplyConfiguration(
+      const clap_plugin_t *plugin,
+      const clap_audio_port_configuration_request *requests,
+      uint32_t request_count) noexcept {
+      auto &self = from(plugin);
+      auto methodName = "clap_plugin_configurable_audio_ports.can_apply_configuration";
+      self.ensureMainThread(methodName);
+      self.ensureIsInactive(methodName);
+      self.ensureClapAudioPortConfigurationRequestIsValid(requests, request_count);
+
+      return self.configurableAudioPortsCanApplyConfiguration(requests, request_count);
+   }
+
+   template <MisbehaviourHandler h, CheckingLevel l>
+   bool Plugin<h, l>::clapConfigurableAudioPortsApplyActivation(
+      const clap_plugin_t *plugin,
+      const clap_audio_port_configuration_request *requests,
+      uint32_t request_count) noexcept {
+      auto &self = from(plugin);
+      auto methodName = "clap_plugin_configurable_audio_ports.apply_configuration";
+      self.ensureMainThread(methodName);
+      self.ensureIsInactive(methodName);
+      self.ensureClapAudioPortConfigurationRequestIsValid(requests, request_count);
+
+      if (l >= CheckingLevel::Minimal &&
+          !self.configurableAudioPortsCanApplyConfiguration(requests, request_count)) {
+         self.hostMisbehaving(
+            "Host requested a configuration that the plugin did not accept. Check with "
+            "clap_plugin_configurable_audio_ports.can_apply_configuration before applying a "
+            "configuration!");
+      }
+
+      return self.configurableAudioPortsApplyConfiguration(requests, request_count);
+   }
+
+   template <MisbehaviourHandler h, CheckingLevel l>
+   void Plugin<h, l>::ensureClapAudioPortConfigurationRequestIsValid(
+      const clap_audio_port_configuration_request *requests,
+      uint32_t request_count) {
+      if (l == CheckingLevel::None)
+         return;
+
+      for (int i = 0; i < request_count; ++i) {
+         if (!requests[i].port_type)
+            continue;
+
+         if (strcmp(requests[i].port_type, CLAP_PORT_MONO) == 0 && requests[i].channel_count != 1) {
+            hostMisbehaving("Host requested a mono port type with a channel count other than one.");
+         } else if (strcmp(requests[i].port_type, CLAP_PORT_STEREO) == 0 && requests[i].channel_count != 2) {
+            hostMisbehaving("Host requested a stereo port type with a channel count other than two.");
+         } else if (strcmp(requests[i].port_type, CLAP_PORT_SURROUND) == 0 && requests[i].channel_count < 3) {
```

Comment:
Our surround representation would allow to have just one channel marked with the semantic sub woofer maybe.
Here I'm not too sure what the correct check would be, I think channel count should just be at least 1 for surround.

The fact that the port is surround lets us associate a channel with its role, but I would not prevent an application to split one port in two different one for processing them differently, and for example processing the bass with one specific plugin while processing the rest with another then merging the results.

If the plugin doesn't allow it, then `can_apply_configuration()` and `apply_configuration()` will simply fail.